File: 02conceitosFundamentais/ordemSuperior.py
from functools import reduce

# Função que adiciona um sufixo " - aprovado" ao nome do aluno
def aprovarAluno(nome):
    return nome + " - aprovado"

# Lista original de alunos
alunos = ["João", "Maria", "Pedro"]

# Usando map para aplicar a função a todos os alunos
alunosAprovados = list(map(aprovarAluno, alunos))

print(alunosAprovados)


# Função que testa se a idade é maior ou igual a 18
def maiorDeIdade(idade):
    return idade >= 18

idades = [12, 17, 25, 30, 14, 23, 8]

# Aplicando filter
maioresDeIdade = list(filter(maiorDeIdade, idades))

print(maioresDeIdade)


# Função que subtrai dois números
def operacao(x, y):
    return x - y

numeros = [1, 2, 3, 4]

# Usando reduce
resultado = reduce(operacao, numeros)

print(resultado)

File: 02conceitosFundamentais/ordemSuperior.hs
-- Função que adiciona " - aprovado" ao nome do aluno
aprovarAluno :: String -> String
aprovarAluno nome = nome ++ " - aprovado"

-- Lista original
alunos :: [String]
alunos = ["Ana", "Maria", "Pedro"]

-- Nova lista usando map
alunosAprovados :: [String]
alunosAprovados = map aprovarAluno alunos


-- Função que verifica se uma idade é maior ou igual a 18
maiorDeIdade :: Int -> Bool
maiorDeIdade idade = idade >= 18

idades :: [Int]
idades = [12, 17, 18, 20, 19, 10, 29]

-- Aplicando filter
maioresDeIdade :: [Int]
maioresDeIdade = filter maiorDeIdade idades


-- Função de subtração
operacao :: Int -> Int -> Int
operacao x y = x + y

numeros :: [Int]
numeros = [1, 2, 3, 4]

-- Usando foldl para reduzir
resultado :: Int
resultado = foldl operacao 0 numeros


main = do
    print alunosAprovados
    print maioresDeIdade
    print resultado